File: eummd/R/naive.R
#' Naive computation for Maximum Mean Discrepancy
#'
#' Computes maximum mean discrepancy statistics with Laplacian 
#' or Gaussian kernel. 
#' Suitable for multivariate data. Naive approach, quadratic in number
#' of observations.
#' 
#' @param X Matrix (or vector) of observations in first sample.
#' 
#' @param Y Matrix (or vector) of observations in second sample.
#' 
#' @param beta kernel parameter. Must be positive; if not, computes
#'             median heuristic in quadratic time. Default value
#'             is \code{-0.1}, which will force median heuristic to be used.
#' 
#' @param pval Boolean for whether to compute p-value or not. 
#' 
#' @param kernel String, either \code{"Laplacian"} or \code{"Gaussian"}. 
#'               Default is \code{"Laplacian"}.
#'
#' @param numperm Number of permutations. Default is \code{200}.
#'
#' @param seednum Seed number for generating permutations. Default is \code{0}, 
#'                which means seed is set randomly. For values larger than 
#'                \code{0}, results will be reproducible.
#' 
#' @param alternative A character string specifying the alternative hypothesis,
#'                    which must be either \code{"greater"} (default) or
#'                    \code{"two.sided"}. In Gretton et al., the 
#'                    MMD test statistic is specified so that if it is 
#'                    significantly larger than zero, then the null hypothesis
#'                    that the two samples come from the same distribution 
#'                    should be rejected. For this reason, \code{"greater"}
#'                    is recommended. The test will still work 
#'                    in many cases with \code{"two.sided"} specified, but this
#'                    could lead to problems in certain cases.
#'
#' @param allowzeropval A boolean, specifying whether we will allow zero 
#'                      p-values or not. Default is \code{FALSE}; then 
#'                      a threshold of \code{0.5 / (numperm+1)} is used, 
#'                      and if the computed p-value is less than this
#'                      threshold, it is then set to be this value.
#'                      this avoids the possibility of zero p-values.
#'
#' @details First checks number of columns (dimension) are equal. 
#'          Suppose matrix \eqn{X} has \eqn{n} rows and \eqn{d} columns, 
#'          and matrix \eqn{Y} has \eqn{m} rows; checks that \eqn{Y} 
#'          has \eqn{d} columns (if not, then throws error). 
#'          Then flattens matrices to vectors (or, if \eqn{d=1}, they are
#'          already vectors.
#'          Then calls C++ method. If the first sample has \eqn{n} 
#'          \eqn{d}-dimensional samples and the second sample has 
#'          \eqn{m} \eqn{d}-dimensional samples, then the algorithm
#'          computes the statistic in \eqn{O((n+m)^2)} time.
#'          
#'  Median difference is as follows:
#'          
#'  \deqn{ m = \textnormal{median} \{ || x_i - x_j ||_1; \,\, i>j, \,\, 
#'         i=1, 2,\dots, n+m,\,\,\textnormal{ and } j=1, 2,\dots, i-1 \}, }
#'    
#'          
#'  where \eqn{ || x_i - x_j ||_1} is the 1-norm, and so if the data 
#'  are \eqn{d}-dimensional then
#'          
#'  \deqn{ || x_i - x_j ||_1 = \sum_{k=1}^{d} |x_{i,k} - x_{j,k}|, }
#'          
#'  and finally median heuristic is \code{beta = 1/m}.
#'  This can be computed in \eqn{O( (n+m)^2 )} time.
#'          
#'  The Laplacian kernel \eqn{k} is defined as 
#'         
#'  \deqn{ k(x,y) = \exp( -\beta || x_i - x_j ||_1 ). }
#'
#'  Random seed is set for \code{std::mt19937} and \code{std::shuffle} in C++.
#'
#' @return A list with the following elements:
#'         \describe{
#'             \item{\code{pval}}{The p-value of the test, if it is  
#'                                computed (\code{pval=TRUE}). }
#'             \item{\code{stat}}{The statistic of the test, which
#'                                is always computed. }
#'             \item{\code{beta}}{The kernel parameter used in the test.
#'                                If \code{beta} was not initialised or
#'                                negative, this will be the median heuristic
#'                                value.}
#'          }
#'
#' @references
#'    Gretton, A., Borgwardt, K. M., Rasch M. J., Schölkopf, B. and Smola, A. 
#'    (2012) "A kernel two-sample test." The Journal of Machine Learning Research 
#'     13, no. 1, 723-773.
#'
#' @examples
#'
#' X <- matrix(c(1:12), ncol=2, byrow=TRUE)
#' Y <- matrix(c(13:20), ncol=2, byrow=TRUE)
#' mmdList <- mmd(X=X, Y=Y, beta=0.1, pval=FALSE)
#'
#' #using median heuristic
#' mmdList <- mmd(X=X, Y=Y, pval=FALSE)
#'
#' #using median heuristic and computing p-value
#' mmdList <- mmd(X=X, Y=Y)
#' \donttest{
#' #using median heuristic and computing p-value
#' #using 1000 permutations and seed 1 for reproducibility.
#' mmdList <- mmd(X=X, Y=Y, numperm=1000, seednum=1)
#' }
#'
#' @export
mmd <- function(X, Y, beta=-0.1, pval=TRUE, kernel=c("Laplacian", "Gaussian"), 
                numperm=200, seednum=0, alternative=c("greater", "two.sided"),
                allowzeropval=FALSE){

    # check vectors/matrices are numeric
    if ( !(is.numeric(X)) ){
        stop("X needs to be numeric.")
    }
    if ( !(is.numeric(Y)) ){
        stop("Y needs to be numeric.")
    }

    # check kernel is correct
    kernel <- kernel[1]
    if ( (kernel != "Laplacian") && (kernel != "Gaussian") ){
        stop("kernel needs to be either 'Laplacian' or 'Gaussian'.")
    }

    # if alternative is 'greater', default
    alternative <- alternative[1]
    if ( (alternative != "greater") && (alternative != "two.sided") ){
        stop("alternative needs to be either 'greater' or 'two.sided'.")
    }

    # two sided is false by default; false is 0
    twosided <- 0
    if (alternative == "two.sided")
        twosided <- 1

    # boundminpval is true by default; true is 1, false is 0
    boundminpval <- 1
    if (allowzeropval==TRUE)
        boundminpval <- 0

    # initialise, will update later
    nX <- 0
    dX <- 0
    nY <- 0
    dY <- 0
    Xvec <- c()
    Yvec <- c()

    # add checks for matrix/vectors
    if (is.matrix(X)){
        # get dimensions of matrices
        nX <- dim(X)[1]
        dX <- dim(X)[2]

        if (!(is.matrix(Y))){
            stop("If X is a matrix, Y must also be a matrix.")
        }

        nY <- dim(Y)[1]
        dY <- dim(Y)[2]

        # check dimensions compatible
        if (dX != dY){
            stop("Dimension (number of columns) of matrices need to be equal.")
        }

        # flatten to vectors; we use transpose
        # > X
        #      [,1] [,2] [,3]
        # [1,]    1    3    5
        # [2,]    2    4    6
        #
        # will be flattened to 
        # [1] 1 3 5 2 4 6
        # which is what we want

        Xvec <- as.vector(t(X))
        Yvec <- as.vector(t(Y))
    } else if (is.vector(X)){
        if (!(is.vector(Y))){
            stop("If X is a vector, Y must also be a vector.")
        }
        nX <- length(X)
        dX <- 1
        nY <- length(Y)
        dY <- 1
        Xvec <- X
        Yvec <- Y
    } else {
        stop("X must be a vector or a matrix.")
    }



    # if beta not positive, compute median heuristic (also from C++)
    # finally, compute MMD 
    mmdList <- list()
    if (kernel=="Gaussian"){
        if (pval){
            mmdList <- mmd_gau_pval_Rcpp(Xvec, Yvec, nX, dX, nY, dY, 
                                         numperm, seednum, beta, twosided, 
                                         boundminpval)
        } else {
            mmdList <- mmd_gau_Rcpp(Xvec, Yvec, nX, dX, nY, dY, beta)
        }
    } else {
        if (pval){
            mmdList <- mmd_lap_pval_Rcpp(Xvec, Yvec, nX, dX, nY, dY, 
                                         numperm, seednum, beta, twosided, 
                                         boundminpval)
        } else {
            mmdList <- mmd_lap_Rcpp(Xvec, Yvec, nX, dX, nY, dY, beta)
        }
    }

    # check pval; if no pval, functions return -1, then changes to NA
    if (mmdList$pval < 0){
        mmdList$pval <- NA
    }
    return(mmdList)
}

File: eummd/src/naive.cpp
// for std::abs
// #include<cmath>
// #include<math.h>
#include<vector>
#include<random>
#include<math.h>
#include<algorithm>


/* Naive MMD with Laplacian kernel
 *
 * Compute the MMD for two matrices
 *
 * @param X First matrix, double*
 *
 * @param Y Second matrix, double*
 *
 * @param beta Parameter for Laplacian kernel.
 *
 * @return A scalar, the kernel value of the two vectors.
 */

double cpp_mmd_lap(double* X, double* Y, 
                   int nX, int dX,
                   int nY, int dY,
                   double beta){


    if (dX != dY){
        //error, dimensions incorrect
        //throw std::invalid_argument("Invalid dimensions for X and Y.");
        return -1.0;
    }

    double arg=0;
    //terms used in computing MMD
    double T1 = 0;
    for (int i = 0; i < nX; ++i){
        for (int j = 0; j < nX; ++j){
            if (i != j){
                arg = 0;
                for (int alpha=0; alpha < dX; ++alpha){
                    arg += std::abs( X[i*dX+alpha] - X[j*dX+alpha] ); 
                }
                T1 += std::exp( -beta * arg );
            }
        }
    }

    double T2 = 0;
    for (int p = 0; p < nY; ++p){
        for (int q = 0; q < nY; ++q){
            if (p != q){
                arg = 0;
                for (int alpha=0; alpha < dY; ++alpha){
                    arg += std::abs( Y[p*dY+alpha] - Y[q*dY+alpha] ); 
                }
                T2 += std::exp( -beta * arg );
            }
        }
    }

    double T3 = 0;
    for (int i = 0; i < nX; ++i){
        for (int p = 0; p < nY; ++p){
            arg = 0;
            for (int alpha=0; alpha < dY; ++alpha){
                arg += std::abs( X[i*dX+alpha] - Y[p*dY+alpha] );
            }
            T3 += std::exp(  -beta * arg  );
        }
    }

    double c1 = 0;
    if (nX > 1){
        c1 = 1.0 / (nX * (nX-1) );
    }
    double c2 = 0;
    if (nY > 1){
        c2 = 1.0 / (nY * (nY-1) );
    }
    double c3 = -2.0 / (nX * nY);

    double MMD_u = c1*T1 + c2*T2 + c3*T3 ;
    return MMD_u;
} // end of cpp_mmd_lap



/* MMD with Gaussian kernel
 *
 * Compute the MMD for two matrices
 *
 * @param X First matrix, double*
 *
 * @param Y Second matrix, double*
 *
 * @param beta Parameter for Laplacian kernel.
 *
 * @return A scalar, the kernel value of the two vectors.
 */

double cpp_mmd_gau(double* X, double* Y, 
                   int nX, int dX,
                   int nY, int dY,
                   double beta){


    if (dX != dY){
        //error, dimensions incorrect
        return -1.0;
    }

    // temporary variables; only compute difference and exp once
    double arg=0;
    double diff = 0;
    //terms used in computing MMD
    double T1 = 0;
    for (int i = 0; i < nX; ++i){
        for (int j = 0; j < nX; ++j){
            if (i != j){
                arg = 0;
                for (int alpha=0; alpha < dX; ++alpha){
                    diff = X[i*dX+alpha] - X[j*dX+alpha];
                    arg += diff*diff; 
                }
                T1 += std::exp( -beta * arg );
            }
        }
    }

    double T2 = 0;
    for (int p = 0; p < nY; ++p){
        for (int q = 0; q < nY; ++q){
            if (p != q){
                arg = 0;
                for (int alpha=0; alpha < dY; ++alpha){
                    diff = Y[p*dY+alpha] - Y[q*dY+alpha];
                    arg += diff*diff; 
                }
                T2 += std::exp( -beta * arg );
            }
        }
    }

    double T3 = 0;
    for (int i = 0; i < nX; ++i){
        for (int p = 0; p < nY; ++p){
            arg = 0;
            for (int alpha=0; alpha < dY; ++alpha){
                diff = X[i*dX+alpha] - Y[p*dY+alpha];
                arg += diff*diff;
            }
            T3 += std::exp( -beta * arg  );
        }
    }

    double c1 = 0;
    if (nX > 1){
        c1 = 1.0 / (nX * (nX-1) );
    }
    double c2 = 0;
    if (nY > 1){
        c2 = 1.0 / (nY * (nY-1) );
    }
    double c3 = -2.0 / (nX * nY);

    double MMD_u = c1*T1 + c2*T2 + c3*T3 ;
    return MMD_u;
}
// end of cpp_mmd_gau



// using only a single vector of values and a single index vector, 
// and noting the sizes of X and Y, and the dimension d
// uses Zindex to extract rows; used for shuffling
double cpp_mmd_lap_ptr(std::vector<double> &Z, 
                       std::vector<size_t> &Zindex,
                       size_t nX,
                       size_t nY,
                       size_t dZ,
                       double beta){

    double arg=0;
    size_t nZ = nX + nY;

    //terms used in computing MMD
    double T1 = 0;
    for (size_t i = 0; i < nX; ++i){
        for (size_t j = 0; j < nX; ++j){
            if (i != j){
                arg = 0;
                for (size_t alpha=0; alpha < dZ; ++alpha){
                    arg += std::abs( Z.at(Zindex.at(i)*dZ + alpha) - Z.at(Zindex.at(j)*dZ + alpha) ); 
                }
                T1 += std::exp( -beta * arg );
            }
        }
    }

    double T2 = 0;
    for (size_t p = nX; p < nZ; ++p){
        for (size_t q = nX; q < nZ; ++q){
            if (p != q){
                arg = 0;
                for (size_t alpha=0; alpha < dZ; ++alpha){
                    arg += std::abs( Z.at( Zindex.at(p)*dZ+alpha ) - Z.at( Zindex.at(q)*dZ+alpha ) ); 
                }
                T2 += std::exp( -beta * arg );
            }
        }
    }

    double T3 = 0;
    for (size_t i = 0; i < nX; ++i){
        for (size_t p = nX; p < nZ; ++p){
            arg = 0;
            for (size_t alpha=0; alpha < dZ; ++alpha){
                arg += std::abs( Z.at(Zindex.at(i)*dZ+alpha) - Z.at(Zindex.at(p)*dZ+alpha) );
            }
            T3 += std::exp(  -beta * arg  );
        }
    }

    double c1 = 0;
    if (nX > 1){
        c1 = 1.0 / (nX * (nX-1) );
    }
    double c2 = 0;
    if (nY > 1){
        c2 = 1.0 / (nY * (nY-1) );
    }
    double c3 = -2.0 / (nX * nY);

    double MMD_u = c1*T1 + c2*T2 + c3*T3 ;
    return MMD_u;
}
//cpp_mmd_lap_ptr



// return both pval and statistic
std::vector<double> cpp_mmd_lap_pval(double* X, double* Y, 
                                     int nX, int dX,
                                     int nY, int dY,
                                     int numperm, int seednum, double beta, 
                                     int twosided, int boundedminpval){

    // return vector is first pval, then statistic
    std::vector<double> returnvec;

    if (dX != dY){
        //error, dimensions incorrect
        returnvec.push_back(-2.0); 
        returnvec.push_back(-2.0);
        return returnvec;
    }


    //int should really be size_t
    size_t Xsize = nX*dX;
    size_t Ysize = nY*dY;
    size_t Zsize = Xsize + Ysize;

    // create vector
    std::vector<double> Z;
    Z.reserve(Zsize);
    //insert X then Y
    Z.insert(Z.end(), &X[0], &X[0]+Xsize);
    Z.insert(Z.end(), &Y[0], &Y[0]+Ysize);

    //now create indices
    std::vector<size_t> Zindex(nX + nY);
    std::iota(Zindex.begin(), Zindex.end(), 0);

    //compute MMD statistic
    double MMDstar = cpp_mmd_lap_ptr(Z, Zindex, nX, nY, dX, beta);

    // initiate random generator
    std::random_device rd;
    std::mt19937 g;
    // set the seed, if greater than 0
    if (seednum > 0){
        g.seed(seednum);
    } else {
        g.seed(rd());
    }

    double MMDperm = 0.0;
    int MMD_count_below = 1;

    //run the permutations
    for (int i=0; i < numperm; ++i){
        //shuffle
        std::shuffle(Zindex.begin(), Zindex.end(), g);
        MMDperm = cpp_mmd_lap_ptr(Z, Zindex, nX, nY, dX, beta);

        if (MMDperm < MMDstar){
            ++MMD_count_below;
        }
    }

    //now return the threshold
    double pval = MMD_count_below / (numperm + 1.0);

    //make one-sided
    //twosided = 0 means false
    //twosided = 1 means true
    if (twosided==1){
        //make two sided
        pval =  1 - std::abs(1 - 2*pval) ;
    } else {
        //one sided; counting below, so must make 
        // large values into small
        pval = 1 - pval;
    }

    // quick check for minimum possible p-value; avoids pval=0
    // 1 / 2(numperm+1)
    if (boundedminpval==1){
        double pmin = 0.5 / (numperm+1.0);
        if (pval < pmin)
            pval = pmin;
    }

    // return vector is first pval, then statistic
    returnvec.push_back(pval);
    returnvec.push_back(MMDstar);
    return returnvec;
}
// cpp_mmd_lap_pval




// using only a single vector of values and a single index vector, 
// and noting the sizes of X and Y, and the dimension d
double cpp_mmd_gau_ptr(std::vector<double> &Z, 
                       std::vector<size_t> &Zindex,
                       size_t nX,
                       size_t nY,
                       size_t dZ,
                       double beta){

    double arg=0;
    double diff = 0;
    size_t nZ = nX + nY;

    //terms used in computing MMD
    double T1 = 0;
    for (size_t i = 0; i < nX; ++i){
        for (size_t j = 0; j < nX; ++j){
            if (i != j){
                arg = 0;
                for (size_t alpha=0; alpha < dZ; ++alpha){
                    diff = Z.at(Zindex.at(i)*dZ + alpha) - Z.at(Zindex.at(j)*dZ + alpha);
                    arg += diff*diff; 
                }
                T1 += std::exp( -beta * arg );
            }
        }
    }

    double T2 = 0;
    for (size_t p = nX; p < nZ; ++p){
        for (size_t q = nX; q < nZ; ++q){
            if (p != q){
                arg = 0;
                for (size_t alpha=0; alpha < dZ; ++alpha){
                    diff = Z.at( Zindex.at(p)*dZ+alpha ) - Z.at( Zindex.at(q)*dZ+alpha ); 
                    arg += diff * diff;
                }
                T2 += std::exp( -beta * arg );
            }
        }
    }

    double T3 = 0;
    for (size_t i = 0; i < nX; ++i){
        for (size_t p = nX; p < nZ; ++p){
            arg = 0;
            for (size_t alpha=0; alpha < dZ; ++alpha){
                diff = Z.at(Zindex.at(i)*dZ+alpha) - Z.at(Zindex.at(p)*dZ+alpha);
                arg += diff * diff;
            }
            T3 += std::exp(  -beta * arg  );
        }
    }

    double c1 = 0;
    if (nX > 1){
        c1 = 1.0 / (nX * (nX-1) );
    }
    double c2 = 0;
    if (nY > 1){
        c2 = 1.0 / (nY * (nY-1) );
    }
    double c3 = -2.0 / (nX * nY);

    double MMD_u = c1*T1 + c2*T2 + c3*T3 ;
    return MMD_u;
}
//cpp_mmd_gau_ptr



// return both pval and statistic
std::vector<double> cpp_mmd_gau_pval(double* X, double* Y, 
                                     int nX, int dX,
                                     int nY, int dY,
                                     int numperm, int seednum, double beta, 
                                     int twosided, int boundedminpval){

    // return vector is first pval, then statistic
    std::vector<double> returnvec;

    if (dX != dY){
        //error, dimensions incorrect
        returnvec.push_back(-2.0); 
        returnvec.push_back(-2.0);
        return returnvec;
    }

    //int should really be size_t
    size_t Xsize = nX*dX;
    size_t Ysize = nY*dY;
    size_t Zsize = Xsize + Ysize;

    // create vector
    std::vector<double> Z;
    Z.reserve(Zsize);
    //insert X then Y
    Z.insert(Z.end(), &X[0], &X[0]+Xsize);
    Z.insert(Z.end(), &Y[0], &Y[0]+Ysize);

    //now create indices
    std::vector<size_t> Zindex(nX + nY);
    std::iota(Zindex.begin(), Zindex.end(), 0);

    //compute MMD statistic
    double MMDstar = cpp_mmd_gau_ptr(Z, Zindex, nX, nY, dX, beta);

    // initiate random generator
    std::random_device rd;
    std::mt19937 g;
    // set the seed, if greater than 0
    if (seednum > 0){
        g.seed(seednum);
    } else {
        g.seed(rd());
    }

    double MMDperm = 0.0;
    int MMD_count_below = 1;

    //run the permutations
    for (int i=0; i < numperm; ++i){
        //shuffle
        std::shuffle(Zindex.begin(), Zindex.end(), g);
        MMDperm = cpp_mmd_gau_ptr(Z, Zindex, nX, nY, dX, beta);

        if (MMDperm < MMDstar){
            ++MMD_count_below;
        }
    }

    //now return the threshold
    double pval = MMD_count_below / (numperm + 1.0);

    //make one-sided
    // pval = convertTwoSidedPvalueToOneSided(pval);
    if (twosided==1){
        //make two sided
        pval =  1 - std::abs(1 - 2*pval) ;
    } else {
        //one sided; counting below, so must make 
        // large values into small
        pval = 1 - pval;
    }

    // quick check for minimum possible p-value; avoids pval=0
    // 1 / 2(numperm+1)
    if (boundedminpval==1){
        double pmin = 0.5 / (numperm+1.0);
        if (pval < pmin)
            pval = pmin;
    }

    // return vector is first pval, then statistic
    returnvec.push_back(pval);
    returnvec.push_back(MMDstar);
    return returnvec;
}
//cpp_mmd_gau_pval


// Adding energy distance
double cpp_energydist(double* X, double* Y, 
                      int nX, int dX,
                      int nY, int dY){


    if (dX != dY){
        //error, dimensions incorrect
        //throw std::invalid_argument("Invalid dimensions for X and Y.");
        return -1.0;
    }

    double arg=0;
    double diff=0;
    //terms used in computing MMD
    double T1 = 0;
    for (int i = 0; i < nX; ++i){
        for (int j = 0; j < nX; ++j){
            if (i != j){
                arg = 0;
                for (int alpha=0; alpha < dX; ++alpha){
                    diff = X[i*dX+alpha] - X[j*dX+alpha]; 
                    arg += diff*diff;
                }
                T1 += ( -std::sqrt( arg ) );
            }
        }
    }

    double T2 = 0;
    for (int p = 0; p < nY; ++p){
        for (int q = 0; q < nY; ++q){
            if (p != q){
                arg = 0;
                for (int alpha=0; alpha < dY; ++alpha){
                    diff = Y[p*dY+alpha] - Y[q*dY+alpha]; 
                    arg += diff*diff;
                }
                T2 += ( -std::sqrt( arg ) );
            }
        }
    }

    double T3 = 0;
    for (int i = 0; i < nX; ++i){
        for (int p = 0; p < nY; ++p){
            arg = 0;
            for (int alpha=0; alpha < dY; ++alpha){
                diff = X[i*dX+alpha] - Y[p*dY+alpha];
                arg += diff*diff;
            }
            T3 += ( -std::sqrt( arg ) );
        }
    }

    double c1 = 0;
    if (nX > 1){
        c1 = 1.0 / (nX * (nX-1) );
    }
    double c2 = 0;
    if (nY > 1){
        c2 = 1.0 / (nY * (nY-1) );
    }
    double c3 = -2.0 / (nX * nY);

    double energydist = c1*T1 + c2*T2 + c3*T3 ;
    return energydist;

} // end of cpp_energydist


// using only a single vector of values and a single index vector, 
// and noting the sizes of X and Y, and the dimension d
double cpp_energy_ptr(std::vector<double> &Z, 
                       std::vector<size_t> &Zindex,
                       size_t nX,
                       size_t nY,
                       size_t dZ){

    double arg=0;
    double diff = 0;
    size_t nZ = nX + nY;
    //difference between Gaussian kernel and Energy distance:
    // std::exp( -beta * arg ) --> ( -std::sqrt( arg ) )


    //terms used in computing MMD
    double T1 = 0;
    for (size_t i = 0; i < nX; ++i){
        for (size_t j = 0; j < nX; ++j){
            if (i != j){
                arg = 0;
                for (size_t alpha=0; alpha < dZ; ++alpha){
                    diff = Z.at(Zindex.at(i)*dZ + alpha) - Z.at(Zindex.at(j)*dZ + alpha);
                    arg += diff*diff; 
                }
                T1 += ( -std::sqrt( arg ) );
            }
        }
    }

    double T2 = 0;
    for (size_t p = nX; p < nZ; ++p){
        for (size_t q = nX; q < nZ; ++q){
            if (p != q){
                arg = 0;
                for (size_t alpha=0; alpha < dZ; ++alpha){
                    diff = Z.at( Zindex.at(p)*dZ+alpha ) - Z.at( Zindex.at(q)*dZ+alpha ); 
                    arg += diff * diff;
                }
                T2 += ( -std::sqrt( arg ) );
            }
        }
    }

    double T3 = 0;
    for (size_t i = 0; i < nX; ++i){
        for (size_t p = nX; p < nZ; ++p){
            arg = 0;
            for (size_t alpha=0; alpha < dZ; ++alpha){
                diff = Z.at(Zindex.at(i)*dZ+alpha) - Z.at(Zindex.at(p)*dZ+alpha);
                arg += diff * diff;
            }
            T3 += ( -std::sqrt( arg ) );
        }
    }

    double c1 = 0;
    if (nX > 1){
        c1 = 1.0 / (nX * (nX-1) );
    }
    double c2 = 0;
    if (nY > 1){
        c2 = 1.0 / (nY * (nY-1) );
    }
    double c3 = -2.0 / (nX * nY);

    double energydist = c1*T1 + c2*T2 + c3*T3 ;
    return energydist;
}
//cpp_energy_ptr


std::vector<double> cpp_energydist_pval(double* X, double* Y, 
                                        int nX, int dX,
                                        int nY, int dY,
                                        int numperm, int seednum, 
                                        int twosided, int boundedminpval){

    // return vector is first pval, then statistic
    std::vector<double> returnvec;

    if (dX != dY){
        //error, dimensions incorrect
        returnvec.push_back(-2.0); 
        returnvec.push_back(-2.0);
        return returnvec;
    }


    //int should really be size_t
    size_t Xsize = nX*dX;
    size_t Ysize = nY*dY;
    size_t Zsize = Xsize + Ysize;

    // create vector
    std::vector<double> Z;
    Z.reserve(Zsize);
    //insert X then Y
    Z.insert(Z.end(), &X[0], &X[0]+Xsize);
    Z.insert(Z.end(), &Y[0], &Y[0]+Ysize);

    //now create indices
    std::vector<size_t> Zindex(nX + nY);
    std::iota(Zindex.begin(), Zindex.end(), 0);


    //compute energy statistic
    double energystar = cpp_energy_ptr(Z, Zindex, nX, nY, dX);

    // initiate random generator
    std::random_device rd;
    std::mt19937 g;
    // set the seed, if greater than 0
    if (seednum > 0){
        g.seed(seednum);
    } else {
        g.seed(rd());
    }

    double energyperm = 0.0;
    int energy_count_below = 1;

    //run the permutations
    for (int i=0; i < numperm; ++i){
        //shuffle
        std::shuffle(Zindex.begin(), Zindex.end(), g);
        energyperm = cpp_energy_ptr(Z, Zindex, nX, nY, dX);

        if (energyperm < energystar){
            ++energy_count_below;
        }
    }

    //now return the threshold
    double pval = energy_count_below / (numperm + 1.0);

    //make one-sided
    //twosided = 0 means false
    //twosided = 1 means true
    if (twosided==1){
        //make two sided
        pval =  1 - std::abs(1 - 2*pval) ;
    } else {
        //one sided; counting below, so must make 
        // large values into small
        pval = 1 - pval;
    }

    // quick check for minimum possible p-value; avoids pval=0
    // 1 / 2(numperm+1)
    if (boundedminpval==1){
        double pmin = 0.5 / (numperm+1.0);
        if (pval < pmin)
            pval = pmin;
    }

    // return vector is first pval, then statistic
    returnvec.push_back(pval);
    returnvec.push_back(energystar);
    return returnvec;
}
